portal-impl/src/com/liferay/portal/db/partition/DBPartitionUtil.java

```diff
@@ -184,12 +200,14 @@ public static void setDefaultCompanyId(long companyId) {
 		}
 	}
 
-	public static BaseModel<?> toEntityModel(BaseModel<?> entityModel) {
+	public static BaseModel<?> toEntityModel(
+		Class<?> entityClass, BaseModel<?> entityModel) {
+
 		if (!_DATABASE_PARTITION_ENABLED) {
 			return entityModel;
 		}
 
-		if (entityModel instanceof ShardedModel) {
+		if (entityClass.isAssignableFrom(ShardedModel.class)) {
```

Comment:
Why you have to pass in "Class<?> entityClass" to change the instanceof to isAssignableFrom ? The casting is done to entityModel instance, it is cheaper to purely operate around it.

Reply:
That's part of the reason why I did not let it returning NullModel. NullModel should be neutral, it should work for any service builder entities. I doubt we have the need to make it a SharedModel.